exposed-core/src/main/kotlin/org/jetbrains/exposed/sql/statements/UpsertStatement.kt

```diff
@@ -61,3 +69,63 @@ open class UpsertStatement<Key : Any>(
         return super.prepared(transaction, sql)
     }
 }
+
+/**
+ * Common interface for building SQL statements that either insert a new row into a table,
+ * or update the existing row if insertion would violate a unique constraint.
+ */
+internal interface UpsertBuilder {
+    /** Returns the columns to be used in the conflict condition of an upsert statement. */
+    fun getKeyColumns(table: Table, vararg keys: Column<*>): List<Column<*>> {
+        return keys.toList().ifEmpty {
+            table.primaryKey?.columns?.toList() ?: table.indices.firstOrNull { it.unique }?.columns
+        } ?: emptyList()
+    }
+
+    /** Returns the columns to be used in the update clause of an upsert statement, along with their insert column reference. */
+    fun getUpdateColumns(
+        dataColumns: List<Column<*>>,
+        toExclude: List<Column<*>>?,
+        keyColumns: List<Column<*>>?
+    ): List<Pair<Column<*>, Expression<*>>> {
+        val updateColumns = toExclude?.let { dataColumns - it.toSet() } ?: dataColumns
+        val updateColumnsWithoutKeys = keyColumns?.let { keys ->
+            updateColumns.filter { it !in keys }.ifEmpty { updateColumns }
+        } ?: updateColumns
+        return updateColumnsWithoutKeys.zip(updateColumnsWithoutKeys.map { it.asForInsert() })
+    }
+
+    /**
+     * Specifies that this column should be updated using the same values that would be inserted if there was
+     * no violation of a unique constraint in an upsert statement.
+     *
+     * @sample org.jetbrains.exposed.sql.tests.shared.dml.UpsertTests.testUpsertWithManualUpdateUsingInsertValues
+     */
+    fun <T> Column<T>.asForInsert(): ExpressionWithColumnType<T> = AsForInsert(this, this.columnType)
```

Comment:
I'm always open to naming suggestions, here and elsewhere.

Most other ORMs support this syntax by calling it `excluded`, though I'm hesitant to do that because it's a strong bias towards PostgreSQL.
For example, in ORMs that chain for statement building, the pattern is usually something like:
```
upsert().set { excluded -> column = excluded.column }

// or

upsert().set(column to excluded(column))
```

This implementation could also be done as a regular function, instead of an extension function, if preferred.

Reply:
@e5l @obabichevjb  This expression will generate 4 different syntax depending on the db:
`EXCLUDED.column`, `NEW.column`, `VALUES(column)`, or `S.column`.

I find `Table.column.excluded()` quite unclear, especially if not coming from a PostgreSQL background knowing exactly the SQL needed.
```kt
TestTable.upsert(
    onUpdate = { listOf(TestTable.word to TestTable.word.excluded()) }
)
```

`Table.column.new()` or `new(Table.column)` may a bit more clear and I like it, but it is still a MySQL-specific syntax. Although, MySQL would accept `EXCLUDED` (or any alias) as well, since the syntax is only based on the alias of the insert values (which we chose to set as `NEW`).
```kt
TestTable.upsert(
    onUpdate = { listOf(TestTable.word to new(TestTable.word)) }
)
```

Another alternative that we could consider is a variant of: `Table.column.insertValue()`, `insertValue(Table.column)`, `insertValueFor(Table.column)`, `insertValueOf(Table.column)`.
```kt
TestTable.upsert(
    onUpdate = { listOf(TestTable.word to TestTable.word.insertValue()) }
)
```

Reply:
I had a draft here and GitHub removed it =( never trust browser inputs...

I had a thought that it could be better to make a function in the context of `onUpdate` with column argument instead of column extension function.

In this case it would be easier to find it in this context. If you know that you expect `VALUES(column)` in sql, you can start to type `values` word and get autocomplete of function with one column argument, and it will be clear what it does.

I don't like when multiple functions with different names do the same thing, but we may provide two function `excluded(column)` and `values(column)`. More generic variant I can imagine now are `fromInsert()`, `insertValue()`.

One another variant, I believe less, because it could be harder to implement and maintain is providing a constant (let's say `EXCLUDED`) that would have same type as original table. In this case it would be possible to write something like `TestTable.word to EXCLUDED.word`

Reply:
In general, I have no strong preference here, but it looks like jooq and ktorm use `excluded`, so probably it would be better to align it with them, to make moving from those libraries (and vice versa >_<) eaiser

Reply:
Yup, I agree and I can switch to column argument function instead if we agree on a name.

> In general, I have no strong preference here, but it looks like jooq and ktorm use excluded, so probably it would be better to align it with them, to make moving from those libraries (and vice versa >_<) eaiser

In this case, I don't see it this way, for 2 reasons:
1. I agree it's good to align with other ORM, especially when it makes sense and the common consensus is clear and fits standard SQL (and Kotlin conventions). I don't think there's anything wrong with being different sometimes (or even setting a new standard for the others ourselves), especially if it aligns with the project's goals.
2. They only use `excluded` in very specific context. jooq for example splits its upsert functionality based on database support by using different extension functions on the common insert statement builder. So only a subset of databases will use `.onConflict()` and only those will have access to `excluded()`. Other databases go the way of `.onDuplicateKeyUpdate()` and don't have any access to `excluded()`. ktorm follows similar principle with its database-specific modules and builder classes, so the class that provides `excluded` isn't available in any module but PG and SQLite.

I'm hoping we can come up with a generic variant to cover all db, but about having multiple functions with different names that do the same thing:
- I agree about not liking it, but these functions would only exist within the upsert interface if that helps.

I like `new()` or `insertValue()` but I'll try to think of others.